lib/nimble_options.ex

```diff
@@ -440,6 +442,48 @@ defmodule NimbleOptions do
     end
   end
 
+  defp validate_type({:list, type}, key, values) when is_list(values) do
+    validation =
+      Enum.reduce(values, [valid_values: [], custom_errors: [], is_valid?: true], fn value, acc ->
+        case validate_type(type, key, value) do
+          :ok ->
+            Keyword.update!(acc, :valid_values, &[value | &1])
+
+          {:ok, value} ->
+            Keyword.update!(acc, :valid_values, &[value | &1])
+
+          {:error, msg} ->
+            acc
+            |> Keyword.put(:is_valid?, false)
+            |> Keyword.update!(:custom_errors, &[msg | &1])
+        end
+      end)
+
+    case Keyword.get(validation, :is_valid?) do
+      true ->
+        valid = validation |> Keyword.get(:valid_values) |> Enum.reverse()
+        {:ok, valid}
+
+      false ->
+        if match?({:custom, _, _, _}, type) do
+          custom = validation |> Keyword.get(:custom_errors) |> Enum.reverse()
+          {:error, custom}
+        else
+          msg =
+            "expected #{inspect(key)} to be a list of type #{inspect(type)}, got: #{
+              inspect(values)
+            }"
+
+          {:error, msg}
+        end
+    end
+  end
+
+  defp validate_type({:list, type}, key, value) do
+    {:error,
+     "expected #{inspect(key)} to be a list of type #{inspect(type)}, got: #{inspect(value)}"}
+  end
```

Comment:
To the end user, this might look really weird. Imagine this:

```
expected :users to be a list of type {:custom, MyModule, :validate_user, []}, got: :foo
```

What is `{:custom, MyModule, :validate_user, []}`? It's confusing IMO.

Let's go with something like this:

```suggestion
  defp validate_type({:list, _type}, key, value) do
    {:error, "expected #{inspect(key)} to be a list, got: #{inspect(value)}"}
  end
```

Reply:
Changed to that simpler error message when list is composed of custom types, see: https://github.com/dashbitco/nimble_options/pull/35/commits/19b9366454a83ff2170404a8c8e19c4b2c1eed73#diff-6c7b866f243a12c3add997a69689b569R467

But kept a more detailed error message if the list is composed of basic types, see: https://github.com/dashbitco/nimble_options/pull/35/commits/19b9366454a83ff2170404a8c8e19c4b2c1eed73#diff-6c7b866f243a12c3add997a69689b569R469